### ISSUE TITLE
InnerTypeNumber

### ISSUE THREAD

**Issue by xenoterracide** [checkstyle/checkstyle#14694]
To complement https://checkstyle.org/checks/sizes/outertypenumber.html

| name	     |  description	                                   | type	| default |
|------------|-----------------------------------------------------|------------|---------|
| max |	Specify the maximum number of outer types allowed. |	[int]	| 2 |

OK. No inner types
```java
class Foo {
}
```

Ok. 1 inner type

```java
class Foo {
  class Bar {
  }
}
```

Ok. 2 inner types

```java
class Foo {
   interface Bar {}
   record Baz {}
}
```

Violation. Things should be moved to an upper level
```java
class Foo {
   class Blah {}
   interface Bar {}
   record Baz {}
}
```

**Comment by romani** [comment]
I am ok with this proposal.
@rnveach , please review.

**Comment by rnveach** [comment]
Would about inner inner types? Would this combine with the inner type count, or would it restart over again, or would they be ignored?

**Comment by romani** [comment]
One more question, should we count inner types in inner types? 
I think better to skip, but let share ideas.

**Comment by xenoterracide** [comment]
I would probably fail this when set to 2. whilst code this simple is readable, in projects over 100k lines... anything doing that is going to become hard to follow, and the point of setting it is another way of looking at code complexity.

**Comment by romani** [comment]
Lets ignore inner in inner for first version of this Check.
We can always extend it to count inner in inner by special property.

Inner in inner will disappear as inner is moved out.

Nuance is with visibility of inner types, it is good to have bunch of them as private, so keep all guts inside.

Public inner types is just laziness of engineers to refactor more as inner type become more mature.

Do you agree to limit Check for public types only? Or to has `publicMax`, `pxxxxxMax` properties?

**Comment by xenoterracide** [comment]
> Do you agree to limit Check for public types only? Or to has publicMax, pxxxxxMax properties?

pxxxMax might be fine. I honestly make most things package protected. Often need them for testing in a more SOLID fashion. Rarely do I have anything other than fields that are private. JPMS is for making package, protected, and public things private to external consumers ;) .
